### PR TITLE
Add service to change camera follow pgain.

### LINKED ISSUE #514: No ability to make follow camera "fixed" or set pgain.
## Description
* Expected behavior: Camera in follow mode allows you to have a nearly fixed view and see vehicle especially at high speeds.
* Actual behavior: When flying planes at 100+ m/s they turn into an invisible spec.

### PR DESCRIPTION
**Pull request by bperseghetti**
# 🦟 Bug fix

Fixes #514

## Summary
Now you can set the pgain of the follow camera when following objects. This is needed for extremely high speed objects.

https://user-images.githubusercontent.com/10233412/215242320-eaa9d430-e853-49c6-a0de-81a69dc6c258.mp4

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by ahcorde** [comment]
@osrf-jenkins run tests please!

File: examples/standalone/scene_provider/README.md [hunk 1]
```diff
@@ -63,3 +63,15 @@ Update follow offset:
 ```
 gz service -s /gui/follow/offset --reqtype gz.msgs.Vector3d --reptype gz.msgs.Boolean --timeout 2000 --req 'x: 5, y: 5, z: 5'
 ```
+
+Set next follow pgain:
+
+```
```
Comment (ahcorde) [inline examples/standalone/scene_provider/README.md:69]:
```suggestion
```bash
```

File: examples/standalone/scene_provider/README.md [hunk 2]
```diff
@@ -63,3 +63,15 @@ Update follow offset:
...
+gz service -s /gui/follow/pgain --reqtype gz.msgs.Double --reptype gz.msgs.Boolean --timeout 2000 --req 'data: 1.0'
+```
+
+Update follow offset with new pgain:
+
+```
```
Comment (ahcorde) [inline examples/standalone/scene_provider/README.md:75]:
```suggestion
```bash
```

File: src/plugins/follow_config/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+gz_gui_add_plugin(FollowConfig
+  SOURCES
+    FollowConfig.cc
+  QT_HEADERS
+    FollowConfig.hh
+  TEST_SOURCES
+    # FollowConfig_TEST.cc
```
Comment (ahcorde) [inline src/plugins/follow_config/CMakeLists.txt:7]: remove?

File: src/plugins/follow_config/FollowConfig.cc [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+#include "gz/gui/MainWindow.hh"
+
+#include <gz/transport/Node.hh>
+
+#include "FollowConfig.hh"
+
+/// \brief Private data class for FollowConfig
+class gz::gui::plugins::FollowConfigPrivate
+{
+  public: std::string followTargetNameService;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:39]: document variables

File: src/plugins/follow_config/FollowConfig.cc [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  public: math::Vector3d followOffset{math::Vector3d(-5.0, 0.0, 3.0)};
+
+  /// \brief Follow P gain
+  public: double followPGain{0.01};
+
+  public: std::string followTargetName;
+
+  public: transport::Node node;
+
+   public: void UpdateFollowTargetName();
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:55]:
```suggestion
  public: void UpdateFollowTargetName();
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+/////////////////////////////////////////////////
+void FollowConfig::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
+{
+  if (this->title.empty())
+    this->title = "Follow Config";
+
+  // Follow target name service
+  this->dataPtr->followTargetNameService = "/gui/follow";
+  gzmsg << "FollowConfig: Follow target name service on ["
+         << this->dataPtr->followTargetNameService << "]" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:91]:
```suggestion
  gzmsg << "FollowConfig: Follow target name service on ["
        << this->dataPtr->followTargetNameService << "]" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+  // Follow target offset service
+  this->dataPtr->followOffsetService = "/gui/follow/offset";
+  gzmsg << "FollowConfig: Follow offset service on ["
+         << this->dataPtr->followOffsetService << "]" << std::endl;
+
+  // Follow target pgain service
+  this->dataPtr->followPGainService = "/gui/follow/pgain";
+  gzmsg << "FollowConfig: Follow P gain service on ["
+         << this->dataPtr->followPGainService << "]" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:101]:
```suggestion
  this->dataPtr->followOffsetService = "/gui/follow/offset";
  gzmsg << "FollowConfig: Follow offset service on ["
        << this->dataPtr->followOffsetService << "]" << std::endl;

  // Follow target pgain service
  this->dataPtr->followPGainService = "/gui/follow/pgain";
  gzmsg << "FollowConfig: Follow P gain service on ["
        << this->dataPtr->followPGainService << "]" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+
+  // Read configuration
+  if (_pluginElem)
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:105]:
```suggestion

  // Read configuration
  if (_pluginElem)
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+  {
+    if (auto nameElem = _pluginElem->FirstChildElement("follow_target"))
+    {
+      this->dataPtr->followTargetName = nameElem->GetText();
+      gzmsg << "FollowConfig: Loaded follow_target from sdf ["
+         << this->dataPtr->followTargetName << "]" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:111]:
```suggestion
      gzmsg << "FollowConfig: Loaded follow_target from sdf ["
            << this->dataPtr->followTargetName << "]" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+      this->dataPtr->newFollowUpdateTargetName = true;
+    }
+    if (auto offsetElem = _pluginElem->FirstChildElement("follow_offset"))
+    {
+      std::stringstream offsetStr;
+      offsetStr << std::string(offsetElem->GetText());
+      offsetStr >> this->dataPtr->followOffset;
+      gzmsg << "FollowConfig: Loaded follow_offset from sdf ["
+         << this->dataPtr->followOffset << "]" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:120]:
```suggestion
      gzmsg << "FollowConfig: Loaded follow_offset from sdf ["
            << this->dataPtr->followOffset << "]" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+      this->dataPtr->newFollowUpdateOffset = true;
+    }
+    if (auto pGainElem = _pluginElem->FirstChildElement("follow_pgain"))
+    {
+      this->dataPtr->followPGain = std::stod(std::string(pGainElem->GetText()));
+      gzmsg << "FollowConfig: Loaded follow_pgain from sdf ["
+         << this->dataPtr->followPGain << "]" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:127]:
```suggestion
      gzmsg << "FollowConfig: Loaded follow_pgain from sdf ["
            << this->dataPtr->followPGain << "]" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+void FollowConfig::SetFollowOffset(double _x,
+          double _y, double _z)
+{
+  if (!this->dataPtr->newFollowUpdateOffset)
+  {
+    this->dataPtr->followOffset = math::Vector3d(
+      _x, _y, _z);
+      gzmsg << "FollowConfig: SetFollowOffset("
+        << this->dataPtr->followOffset << ")" << std::endl;
+    this->dataPtr->newFollowUpdateOffset = true;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:169]:
```suggestion
    this->dataPtr->followOffset = math::Vector3d(
      _x, _y, _z);
    gzmsg << "FollowConfig: SetFollowOffset("
          << this->dataPtr->followOffset << ")" << std::endl;
    this->dataPtr->newFollowUpdateOffset = true;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+{
+  // Offset
+  std::function<void(const msgs::Boolean &, const bool)> cbName =
+    [&](const msgs::Boolean &/*_rep*/, const bool _resultName)
+  {
+    if (!_resultName) {
+      gzerr << "FollowConfig: Error sending follow target name." << std::endl;
+    } else {
+      gzmsg << "FollowConfig: Request Target Name: "
+        << this->followTargetName << " sent" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:196]:
```suggestion
      gzmsg << "FollowConfig: Request Target Name: "
            << this->followTargetName << " sent" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+{
+  // Offset
+  std::function<void(const msgs::Boolean &, const bool)> cbOffset =
+    [&](const msgs::Boolean &/*_rep*/, const bool _resultOffset)
+  {
+    if (!_resultOffset) {
+      gzerr << "FollowConfig: Error sending follow offset." << std::endl;
+    } else {
+      gzmsg << "FollowConfig: Request Offset: "
+        << this->followOffset << " sent" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:217]:
```suggestion
      gzmsg << "FollowConfig: Request Offset: "
            << this->followOffset << " sent" << std::endl;
```

File: src/plugins/follow_config/FollowConfig.cc [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+{
+  // PGain
+  std::function<void(const msgs::Boolean &, const bool)> cbPGain =
+    [&](const msgs::Boolean &/*_rep*/, const bool _resultPGain)
+  {
+    if (!_resultPGain) {
+      gzerr << "FollowConfig: Error sending follow pgain." << std::endl;
+    } else {
+      gzmsg << "FollowConfig: Request PGain: "
+        << this->followPGain << " sent" << std::endl;
```
Comment (ahcorde) [inline src/plugins/follow_config/FollowConfig.cc:240]:
```suggestion
      gzmsg << "FollowConfig: Request PGain: "
            << this->followPGain << " sent" << std::endl;
```

File: test/integration/follow_config.cc [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+using namespace gui;
+using namespace std::chrono_literals;
+
+/////////////////////////////////////////////////
+TEST(MinimalSceneTest, GZ_UTILS_TEST_ENABLED_ONLY_ON_LINUX(Config))
+{
+  common::Console::SetVerbosity(4);
+
+  Application app(g_argc, g_argv);
+  app.AddPluginPath(std::string(PROJECT_BINARY_PATH) + "/lib");
```
Comment (ahcorde) [inline test/integration/follow_config.cc:59]: `gz::common::joinPath()`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/plugins/camera_tracking/CameraTracking.cc [hunk 17]
```diff
@@ -208,12 +219,19 @@ void CameraTrackingPrivate::Initialize()
...
-          << this->followOffsetService << "]" << std::endl;
+  // follow offset
+  this->followOffsetService = "/gui/follow/offset";
+  this->node.Advertise(this->followOffsetService,
+      &CameraTrackingPrivate::OnFollowOffset, this);
+  gzmsg << "Follow offset service on ["
+         << this->followOffsetService << "]" << std::endl;
+
+  // follow pgain
+  this->followPGainService = "/gui/follow/pgain";
```
Comment (iche033) [inline src/plugins/camera_tracking/CameraTracking.cc:230]: suggestion to change `pgain` to `p_gain`, i.e. `/gui/follow/p_gain`, to be consistent with other places in gz that use this term, e.g. https://github.com/gazebosim/gz-sim/blob/gz-sim7/src/systems/joint_controller/JointController.cc#L126

File: src/plugins/follow_config/FollowConfig.cc [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+/////////////////////////////////////////////////
+void FollowConfigPrivate::UpdateFollowOffset()
+{
+  // Offset
+  std::function<void(const msgs::Boolean &, const bool)> cbOffset =
+    [&](const msgs::Boolean &/*_rep*/, const bool _resultOffset)
+  {
+    if (!_resultOffset) {
```
Comment (iche033) [inline src/plugins/follow_config/FollowConfig.cc:213]: gazebo coding style nitpick: bracket on new line

File: test/integration/follow_config.cc (hunk 16 above)
Comment (bperseghetti) [reply test/integration/follow_config.cc:59]: Hmmmm, from what I see looks like all the other plugins use ```app.AddPluginPath(std::string(PROJECT_BINARY_PATH) + "/lib");``` is there a reason to only make this on different?

File: src/plugins/follow_config/FollowConfig.cc (hunk 4 above)
Comment (bperseghetti) [reply src/plugins/follow_config/FollowConfig.cc:39]: Done.

File: src/plugins/follow_config/CMakeLists.txt (hunk 3 above)
Comment (bperseghetti) [reply src/plugins/follow_config/CMakeLists.txt:7]: Makes sense. Done.

File: src/plugins/camera_tracking/CameraTracking.cc (hunk 17 above)
Comment (bperseghetti) [reply src/plugins/camera_tracking/CameraTracking.cc:230]: Changed it everywhere.

File: src/plugins/follow_config/FollowConfig.cc (hunk 18 above)
Comment (bperseghetti) [reply src/plugins/follow_config/FollowConfig.cc:213]: Done.

File: test/integration/follow_config.cc (hunk 16 above)
Comment (iche033) [reply test/integration/follow_config.cc:59]: I think we just try to use `jointPath` whenever we remember as it's recommended. The `joinPaths` function makes the resulting path more platform independent, i.e. it uses `/` for linux  and `\` for windows

File: test/integration/follow_config.cc [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+  {
+    std::this_thread::sleep_for(100ms);
+    QCoreApplication::processEvents();
+  }
+  EXPECT_LT(sleep, maxSleep);
+
+  EXPECT_NEAR(0.0, abs(poseMsg.mutable_position()->x()), 1.0);
+  EXPECT_NEAR(0.0, abs(poseMsg.mutable_position()->y()), 1.0);
+  EXPECT_NEAR(0.0, abs(poseMsg.mutable_position()->z()), 1.0);
+
```
Comment (ahcorde) [inline test/integration/follow_config.cc:166]:
```suggestion
```

File: test/integration/follow_config.cc (hunk 16 above)
Comment (bperseghetti) [reply test/integration/follow_config.cc]: Sounds good, done.

File: test/integration/follow_config.cc (hunk 19 above)
Comment (bperseghetti) [reply test/integration/follow_config.cc]: Removed.

**Review by ahcorde — APPROVED** [review]

**Review by mjcarroll — APPROVED** [review]

File: src/plugins/follow_config/FollowConfig.cc [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+      gzerr << "FollowConfig: Error sending follow target name." << std::endl;
+    } else {
+      gzmsg << "FollowConfig: Request Target Name: "
+            << this->followTargetName << " sent" << std::endl;
+    }
+  };
+
+  msgs::StringMsg reqName;
+  reqName.set_data(this->followTargetName);
+  node.Request(this->followTargetNameService, reqName, cbName);
```
Comment (azeey) [inline src/plugins/follow_config/FollowConfig.cc:213]: If you're using async service requests, you'll need to synchronize access to all the variables used by the callback.

**Review by azeey — CHANGES_REQUESTED** [review]

File: src/plugins/follow_config/FollowConfig.cc (hunk 20 above)
Comment (azeey) [reply src/plugins/follow_config/FollowConfig.cc:213]: Alternatively, just copy-capture the variable in the lambda. I think this is a lot easier.

Comment (bperseghetti) [reply src/plugins/follow_config/FollowConfig.cc:213]: Will finally get back to looking at this.

Comment (ahcorde) [reply src/plugins/follow_config/FollowConfig.cc:213]: @bperseghetti did you address the feedback ?

**Comment by mjcarroll** [comment]
@bperseghetti I haven't seen any movement on this one and it targets garden branches, so I'm going to remove the beta label for now.

**Comment by caguero** [comment]
@bperseghetti , friendly ping. Could you review the feedback and we'll try to merge this patch?

**Comment by bperseghetti** [comment]
> @bperseghetti , friendly ping. Could you review the feedback and we'll try to merge this patch?

Yeah, will look it over again here this weekend and see if I can't make the requested changes.

**Comment by jennuine** [comment]
@bperseghetti there are CI errors: 

```bash
/home/jenkins/workspace/gz_gui-ci-pr_any-focal-amd64/gz-gui/test/integration/follow_config.cc:162
Expected: (sleep) < (maxSleep), actual: 61 vs 60
```

**Comment by bperseghetti** [comment]
> @bperseghetti there are CI errors:
> 
> ```shell
> /home/jenkins/workspace/gz_gui-ci-pr_any-focal-amd64/gz-gui/test/integration/follow_config.cc:162
> Expected: (sleep) < (maxSleep), actual: 61 vs 60
> ```

I actually was planning to close this and open a new PR that has a more "holistic" approach to controlling the follow camera including what was done in here but more.

I just need to rebase that work on humble as I no longer plan to target it to garden (due to relatively sooner EOL).

**Comment by bperseghetti** [comment]
Closing in favor of PR #619

**Pull request closed without merging**